README.rst
ToDus client
============

.. image:: https://img.shields.io/pypi/v/todus.svg
   :target: https://pypi.org/project/todus

.. image:: https://img.shields.io/pypi/pyversions/todus.svg
   :target: https://pypi.org/project/todus

.. image:: https://pepy.tech/badge/todus
   :target: https://pepy.tech/project/todus

.. image:: https://img.shields.io/pypi/l/todus.svg
   :target: https://pypi.org/project/todus

.. image:: https://github.com/adbenitez/todus/actions/workflows/python-ci.yml/badge.svg
   :target: https://github.com/adbenitez/todus/actions/workflows/python-ci.yml

.. image:: https://img.shields.io/badge/code%20style-black-000000.svg
   :target: https://github.com/psf/black

ToDus client for the command line and API to use in your Python projects.

Install
-------

To install the latest stable version run::

  pip install -U todus

To install the latest stable version and enable the 7z multivolume feature for split uploads::

  pip install -U 'todus[7z]'

To test the unreleased version run::

  pip install todus git+https://github.com/adbenitez/todus

Usage
-----

::

   usage: todus [-h] [-n PHONE-NUMBER] [-v] {login,upload,download,token,accounts} ...

   ToDus Client

   positional arguments:
     {login,upload,download,token,accounts}
       login               authenticate in server
       upload              upload file
       download            download file
       token               get a token
       accounts            list accounts

   optional arguments:
     -h, --help            show this help message and exit
     -n PHONE-NUMBER, --number PHONE-NUMBER
                           account's phone number, if not given the default account will be used
     -v, --version         show program's version number and exit.


Developer Quickstart
--------------------

This is an example of how you would use this library in your projects:

.. code-block:: python

  from todus.client import ToDusClient2

  client = ToDusClient2(phone_number="5312345678")

  # this only needs to be done once:
  client.request_code()  # request SMS to start a new session
  pin = input("Enter PIN:").strip()  # enter the PIN received in SMS
  client.validate_code(pin)
  # you must save your session's password to avoid having to verify via SMS again.
  print(f"Save your password: {client.password}")

  # you need to login to upload/download files:
  client.login()

  # uploading a file:
  file_path = "/home/user/Pictures/photo.jpg"
  with open(file_path, "rb") as file:
      data = file.read()
  url = client.upload_file(data)
  print(f"Uploaded file to: {url}")

  # downloading a file:
  size = client.download_file(url, path="my-photo.jpg")
  print(f"Downloaded {size:,} Bytes")


CHANGELOG.rst
Changelog
=========

`Unreleased`_
-------------

- ``todus.errors.AuthenticationError`` is now raised when account fails to login, ``todus.errors.TokenExpiredError`` will be raised when token expired.
- CLI: print program usage if no sub-command is provided.
- CLI: if session expired account password is deleted (session is logged out)
- CLI: added ``--remove`` and ``--default`` options to ``accounts`` subcomand.

`1.1.0`_
--------

- ``py7zr`` is now an optional dependency, ``zipfile`` will be used if ``py7zr`` is not installed.

`1.0.0`_
--------

- added ``todus.util.normalize_phone_number()`` utility.
- added ``todus.client.ToDusClient2`` class.
- removed ``todus.util.ResultProcess`` and ``todus.errors.AbortError`` classes.
- avoid incomplete file downloads, check file size match ``Content-Length``.
- ``todus.client.ToDusClient.download_file()`` will now resume partial downloads.
- added logger parameter to ``todus.client.ToDusClient``.
- updated fake client version.
- CLI: now links can be downloaded concurrently.
- CLI: register ``todus`` command, should be available in the shell.
- CLI: added progress bar.
- CLI: fixed bug that prevented from using "login" subcommand.
- CLI: skip already downloaded/uploaded files.
- CLI: renamed ``--part-size`` option to ``--split``.
- CLI: removed ``--config-folder`` option.
- CLI: save logs in ``~/.todus/``.
- CLI: terminate program gracefully on ``Ctl+C``.
- CLI: improved account management, now logins are saved in ``~/.todus/config.json``.
- CLI: added ``token`` subcommand to get a new token that could be used in other ToDus tools.
- CLI: added ``accounts`` subcomand to list added accounts.

0.1.0
-----

- initial release

.. _Unreleased: https://github.com/adbenitez/todus/compare/v1.1.0...HEAD
.. _1.1.0: https://github.com/adbenitez/todus/compare/v1.0.0...v1.1.0
.. _1.0.0: https://github.com/adbenitez/todus/compare/v0.1.0...v1.0.0


todus/s3.py
import json
import re
import socket
import ssl
from base64 import b64decode, b64encode
from contextlib import contextmanager
from threading import Lock
from typing import Generator

from .errors import AuthenticationError, EndOfStreamError
from .util import generate_token

_BUFFERSIZE = 1024 * 1024
_lock = Lock()


@contextmanager
def _get_socket() -> Generator:
    with _lock:
        context = ssl.create_default_context()
        context.check_hostname = False
        _socket = context.wrap_socket(socket.socket(socket.AF_INET))
        _socket.settimeout(15)
        _socket.connect(("im.todus.cu", 1756))
        _socket.send(
            b"<stream:stream xmlns='jc' o='im.todus.cu' xmlns:stream='x1' v='1.0'>"
        )
        with _socket:
            yield _socket


def _negociate_start(
    response: str, ssl_socket: ssl.SSLSocket, authstr: bytes, sid: str
) -> bool:
    if response.startswith(
        "<?xml version='1.0'?><stream:stream i='"
    ) and response.endswith("xmlns:stream='x1' f='im.todus.cu' xmlns='jc'>"):
        return True

    auth_stream = (
        "<stream:features><es xmlns='x2'><e>PLAIN</e><e>X-OAUTH2</e></es>"
        "<register xmlns='http://jabber.org/features/iq-register'/></stream:features>"
    )
    if response == auth_stream:
        ssl_socket.send(b"<ah xmlns='ah:ns' e='PLAIN'>" + authstr + b"</ah>")
        return True

    if response == "<ok xmlns='x2'/>":
        ssl_socket.send(
            b"<stream:stream xmlns='jc' o='im.todus.cu' xmlns:stream='x1' v='1.0'>"
        )
        return True

    if "<stream:features><b1 xmlns='x4'/>" in response:
        ssl_socket.send(f"<iq i='{sid}-1' t='set'><b1 xmlns='x4'></b1></iq>".encode())
        return True

    return False


def _parse_token(token: str) -> tuple:
    phone = json.loads(b64decode(token.split(".")[1]).decode())["username"]
    authstr = b64encode((chr(0) + phone + chr(0) + token).encode("utf-8"))
    return phone, authstr


def reserve_url(token: str, filesize: int) -> tuple:
    """Reserve file URL to upload.

    Returns a tuple with upload and download URLs.
    """
    phone, authstr = _parse_token(token)
    sid = generate_token(5)

    with _get_socket() as ssl_socket:
        while True:
            response = ssl_socket.recv(_BUFFERSIZE).decode()
            if _negociate_start(response, ssl_socket, authstr, sid):
                continue

            if f"t='result' i='{sid}-1'>" in response:
                ssl_socket.send(b"<en xmlns='x7' u='true' max='300'/>")
                ssl_socket.send(
                    (
                        "<iq i='"
                        + sid
                        + "-3' t='get'><query xmlns='todus:purl' type='1' persistent='false' size='"
                        + str(filesize)
                        + "' room=''></query></iq>"
                    ).encode()
                )
                continue

            if response.startswith("<ed u='true' max='300'"):
                ssl_socket.send(("<p i='" + sid + "-4'></p>").encode())
                continue

            if response.startswith("<iq o='" + phone + "@im.todus.cu"):
                match = re.match(r".*put='(.*)' get='(.*)' stat.*", response)
                assert match, f"Unexpected response: {response}"
                up_url = match.group(1).replace("amp;", "")
                down_url = match.group(2)
                return (up_url, down_url)

            if "<not-authorized/>" in response:
                raise AuthenticationError()

            if not response:
                raise EndOfStreamError()


def get_real_url(token: str, url: str) -> str:
    """Get authenticated URL."""
    authstr = _parse_token(token)[1]
    sid = generate_token(5)

    with _get_socket() as ssl_socket:
        while True:
            response = ssl_socket.recv(_BUFFERSIZE).decode()
            if _negociate_start(response, ssl_socket, authstr, sid):
                continue

            if f"t='result' i='{sid}-1'>" in response:
                data = f"<iq i='{sid}-2' t='get'><query xmlns='todus:gurl' url='{url}'></query></iq>"
                ssl_socket.send(data.encode())
                continue

            if f"t='result' i='{sid}-2'>" in response and "status='200'" in response:
                match = re.match(".*du='(.*)' stat.*", response)
                assert match, f"Unexpected response: {response}"
                return match.group(1).replace("amp;", "")

            if "<not-authorized/>" in response:
                raise AuthenticationError()

            if not response:
                raise EndOfStreamError()


todus/client.py
import functools
import json
import logging
import os
import re
import socket
import ssl
import string
import time
from base64 import b64decode, b64encode
from contextlib import contextmanager
from enum import IntEnum
from http.client import IncompleteRead
from threading import Lock
from typing import Callable, Generator

import requests.exceptions

from .errors import AuthenticationError, EndOfStreamError, TokenExpiredError
from .util import generate_token

_BUFFERSIZE = 1024 * 1024


class FileType(IntEnum):
    """ToDus attachment type."""

    FILE = 0
    VOICE = 1
    AUDIO = 2
    VIDEO = 3
    PICTURE = 4
    PROFILE = 5
    PROFILE_THUMBNAIL = 6


class ToDusClient:
    """Class to interact with the ToDus API."""

    def __init__(
        self,
        version_name: str = "0.40.29",
        version_code: str = "21833",
        logger: logging.Logger = logging,  # type: ignore
    ) -> None:
        self.version_name = version_name
        self.version_code = version_code
        self.logger = logger
        self._lock = Lock()

        self.session = requests.Session()
        self.session.headers.update(
            {
                "Accept-Encoding": "gzip",
            }
        )
        self.session.request = functools.partial(_request, self.session.request)  # type: ignore

    @contextmanager
    def _get_socket(self) -> Generator:
        with self._lock:
            context = ssl.create_default_context()
            context.check_hostname = False
            _socket = context.wrap_socket(socket.socket(socket.AF_INET))
            _socket.settimeout(15)
            _socket.connect(("im.todus.cu", 1756))
            _socket.send(
                b"<stream:stream xmlns='jc' o='im.todus.cu' xmlns:stream='x1' v='1.0'>"
            )
            with _socket:
                yield _socket

    def _reserve_url(self, token: str, filesize: int, file_type: FileType) -> tuple:
        phone, authstr = _parse_token(token)
        sid = generate_token(5)

        with self._get_socket() as ssl_socket:
            while True:
                response = ssl_socket.recv(_BUFFERSIZE).decode()
                if _negociate_start(response, ssl_socket, authstr, sid):
                    continue

                if f"t='result' i='{sid}-1'>" in response:
                    ssl_socket.send(b"<en xmlns='x7' u='true' max='300'/>")
                    ssl_socket.send(
                        (
                            "<iq i='"
                            + sid
                            + "-3' t='get'><query xmlns='todus:purl' type='"
                            + str(int(file_type))
                            + "' persistent='false' size='"
                            + str(filesize)
                            + "' room=''></query></iq>"
                        ).encode()
                    )
                    continue

                if response.startswith("<ed u='true' max='300'"):
                    ssl_socket.send(("<p i='" + sid + "-4'></p>").encode())
                    continue

                if response.startswith("<iq o='" + phone + "@im.todus.cu"):
                    match = re.match(r".*put='(.*)' get='(.*)' stat.*", response)
                    assert match, f"Unexpected response: {response}"
                    up_url = match.group(1).replace("amp;", "")
                    down_url = match.group(2)
                    return (up_url, down_url)

                if "<not-authorized/>" in response:
                    raise TokenExpiredError()

                if not response:
                    raise EndOfStreamError()

    def _get_real_url(self, token: str, url: str) -> str:
        authstr = _parse_token(token)[1]
        sid = generate_token(5)

        with self._get_socket() as ssl_socket:
            while True:
                response = ssl_socket.recv(_BUFFERSIZE).decode()
                if _negociate_start(response, ssl_socket, authstr, sid):
                    continue

                if f"t='result' i='{sid}-1'>" in response:
                    data = f"<iq i='{sid}-2' t='get'><query xmlns='todus:gurl' url='{url}'></query></iq>"
                    ssl_socket.send(data.encode())
                    continue

                if (
                    f"t='result' i='{sid}-2'>" in response
                    and "status='200'" in response
                ):
                    match = re.match(".*du='(.*)' stat.*", response)
                    assert match, f"Unexpected response: {response}"
                    return match.group(1).replace("amp;", "")

                if "<not-authorized/>" in response:
                    raise TokenExpiredError()

                if not response:
                    raise EndOfStreamError()

    @property
    def auth_ua(self) -> str:
        """User Agent used for authentication."""
        return f"ToDus {self.version_name} Auth"

    @property
    def upload_ua(self) -> str:
        """User Agent used for uploads."""
        return f"ToDus {self.version_name} HTTP-Upload"

    @property
    def download_ua(self) -> str:
        """User Agent used for downloads."""
        return f"ToDus {self.version_name} HTTP-Download"

    def request_code(self, phone_number: str) -> None:
        """Request server to send verification SMS code."""
        headers = {
            "Host": "auth.todus.cu",
            "User-Agent": self.auth_ua,
            "Content-Type": "application/x-protobuf",
        }
        data = (
            b"\n\n"
            + phone_number.encode()
            + b"\x12\x96\x01"
            + generate_token(150).encode()
        )
        url = "https://auth.todus.cu/v2/auth/users.reserve"
        with self.session.post(url, data=data, headers=headers) as resp:
            resp.raise_for_status()

    def validate_code(self, phone_number: str, code: str) -> str:
        """Validate phone number with received SMS code.

        Returns the account password.
        """
        headers = {
            "Host": "auth.todus.cu",
            "User-Agent": self.auth_ua,
            "Content-Type": "application/x-protobuf",
        }
        data = (
            b"\n\n"
            + phone_number.encode()
            + b"\x12\x96\x01"
            + generate_token(150).encode()
            + b"\x1a\x06"
            + code.encode()
        )
        url = "https://auth.todus.cu/v2/auth/users.register"
        with self.session.post(url, data=data, headers=headers) as resp:
            resp.raise_for_status()
            if b"`" in resp.content:
                index = resp.content.index(b"`") + 1
                return resp.content[index : index + 96].decode()
            return resp.content[5:166].decode()

    def login(self, phone_number: str, password: str) -> str:
        """Login with phone number and password to get an access token."""
        headers = {
            "Host": "auth.todus.cu",
            "user-agent": self.auth_ua,
            "content-type": "application/x-protobuf",
        }
        data = (
            b"\n\n"
            + phone_number.encode()
            + b"\x12\x96\x01"
            + generate_token(150).encode()
            + b"\x12\x60"
            + password.encode()
            + b"\x1a\x05"
            + self.version_code.encode()
        )
        url = "https://auth.todus.cu/v2/auth/token"
        with self.session.post(url, data=data, headers=headers) as resp:
            if resp.status_code == 403:
                raise AuthenticationError()
            resp.raise_for_status()
            token = "".join([c for c in resp.text if c in string.printable])
            return token

    def upload_file(
        self,
        token: str,
        data: bytes,
        size: int = None,
        file_type: FileType = FileType.VOICE,
    ) -> str:
        """Upload data and return the download URL."""
        up_url, down_url = self._reserve_url(token, size or len(data), file_type)
        headers = {
            "User-Agent": self.upload_ua,
            "Authorization": f"Bearer {token}",
        }
        with self.session.put(
            url=up_url,
            data=data,
            headers=headers,
        ) as resp:
            resp.raise_for_status()
        return down_url

    def download_file(self, token: str, url: str, path: str) -> int:
        """Download file URL.

        Returns the file size.
        """
        temp_path = f"{path}.part"
        url = self._get_real_url(token, url)
        headers = {
            "User-Agent": self.download_ua,
            "Authorization": f"Bearer {token}",
        }
        size = -1
        with open(temp_path, "ab") as file:
            pos = file.tell()
            while pos < size or size == -1:
                if pos:
                    headers["Range"] = f"bytes={pos}-"
                try:
                    with self.session.get(
                        url=url, headers=headers, stream=True
                    ) as resp:
                        resp.raise_for_status()
                        size = pos + int(resp.headers["Content-Length"])
                        try:
                            for chunk in resp.iter_content(chunk_size=10):
                                file.write(chunk)
                        except requests.exceptions.ConnectionError as err:
                            self.logger.exception(err)
                            time.sleep(5)
                except IncompleteRead as err:
                    self.logger.exception(err)
                    time.sleep(5)
                except requests.exceptions.ReadTimeout as err:
                    self.logger.exception(err)
                    time.sleep(5)
                pos = file.tell()
        os.rename(temp_path, path)
        return size


class ToDusClient2(ToDusClient):
    """Class to interact with the ToDus API."""

    def __init__(self, phone_number: str, password: str = "", **kwargs) -> None:
        super().__init__(**kwargs)
        self.phone_number = phone_number
        self.password = password
        self.token = ""

    @property
    def registered(self) -> bool:
        """True if this client has a phone_number and password set."""
        return bool(self.phone_number and self.password)

    @property
    def logged(self) -> bool:
        """True if client is logged in."""
        return bool(self.token)

    def request_code(self) -> None:  # noqa
        """Request server to send verification SMS code."""
        super().request_code(self.phone_number)

    def validate_code(self, code: str) -> None:  # noqa
        """Validate phone number with received SMS code.

        Returns the account password.
        """
        self.password = super().validate_code(self.phone_number, code)

    def login(self) -> None:  # noqa
        """Login with phone number and password to get an access token."""
        assert self.password, "Can't login without password"
        self.token = super().login(self.phone_number, self.password)

    def upload_file(  # noqa
        self, data: bytes, size: int = None, file_type: FileType = FileType.VOICE
    ) -> str:
        """Upload data and return the download URL."""
        assert self.token, "Token needed"
        return super().upload_file(self.token, data, size, file_type)

    def download_file(self, url: str, path: str) -> int:  # noqa
        """Download file URL.

        Returns the file size.
        """
        assert self.token, "Token needed"
        return super().download_file(self.token, url, path)


def _request(real_request: Callable, *args, **kwargs) -> requests.Response:
    kwargs.setdefault("timeout", 30)
    resp = real_request(*args, **kwargs)
    if resp.encoding is None:
        # Default Encoding for HTML4 ISO-8859-1 (Latin-1)
        resp.encoding = "latin-1"
    return resp


def _negociate_start(
    response: str, ssl_socket: ssl.SSLSocket, authstr: bytes, sid: str
) -> bool:
    if response.startswith(
        "<?xml version='1.0'?><stream:stream i='"
    ) and response.endswith("xmlns:stream='x1' f='im.todus.cu' xmlns='jc'>"):
        return True

    auth_stream = (
        "<stream:features><es xmlns='x2'><e>PLAIN</e><e>X-OAUTH2</e></es>"
        "<register xmlns='http://jabber.org/features/iq-register'/></stream:features>"
    )
    if response == auth_stream:
        ssl_socket.send(b"<ah xmlns='ah:ns' e='PLAIN'>" + authstr + b"</ah>")
        return True

    if response == "<ok xmlns='x2'/>":
        ssl_socket.send(
            b"<stream:stream xmlns='jc' o='im.todus.cu' xmlns:stream='x1' v='1.0'>"
        )
        return True

    if "<stream:features><b1 xmlns='x4'/>" in response:
        ssl_socket.send(f"<iq i='{sid}-1' t='set'><b1 xmlns='x4'></b1></iq>".encode())
        return True

    return False


def _parse_token(token: str) -> tuple:
    phone = json.loads(b64decode(token.split(".")[1]).decode())["username"]
    authstr = b64encode((chr(0) + phone + chr(0) + token).encode("utf-8"))
    return phone, authstr


todus/main.py
"""CLI program."""
# pylama:ignore=R0912,C901,R0913

import argparse
import functools
import json
import logging.handlers
import os
import time
from concurrent.futures import ThreadPoolExecutor
from tempfile import TemporaryDirectory
from threading import Lock
from typing import TextIO
from urllib.parse import quote_plus, unquote_plus

import multivolumefile
import tqdm

from . import __version__
from .client import ToDusClient2
from .errors import AuthenticationError
from .util import normalize_phone_number

try:
    import py7zr

    ARCHIVE_EXT = "7z"
except ImportError:
    import zipfile

    ARCHIVE_EXT = "zip"


def _get_config() -> dict:
    with open(CONFIG_PATH, encoding="utf-8") as file:
        return json.load(file)


def _save_config(config: dict) -> None:
    with open(CONFIG_PATH, "w", encoding="utf-8") as file:
        return json.dump(config, file)


def _get_logger() -> logging.Logger:
    logger = logging.Logger(__name__.split(".", maxsplit=1)[0])
    logger.parent = None
    formatter = logging.Formatter("%(asctime)s - %(levelname)s - %(message)s")

    log_path = os.path.join(PROGRAM_FOLDER, "log.txt")
    fhandler = logging.handlers.RotatingFileHandler(
        log_path, backupCount=3, maxBytes=1024 ** 2
    )
    fhandler.setLevel(logging.DEBUG)
    fhandler.setFormatter(formatter)
    logger.addHandler(fhandler)

    return logger


def _split_upload(
    client: ToDusClient2, path: str, part_size: int, max_workers: int
) -> str:
    with open(path, "rb") as file:
        data = file.read()
    filename = os.path.basename(path)
    with TemporaryDirectory() as tempdir:
        with multivolumefile.open(
            os.path.join(tempdir, f"{filename}.{ARCHIVE_EXT}"),
            "wb",
            volume=part_size,
        ) as vol:
            if ARCHIVE_EXT == "7z":
                with py7zr.SevenZipFile(vol, "w") as archive:
                    archive.writestr(data, filename)
            else:
                with zipfile.ZipFile(vol, "w", zipfile.ZIP_DEFLATED) as archive:  # type: ignore
                    archive.writestr(filename, data)
        del data
        path = os.path.abspath(filename + ".txt")
        uploaded_parts = []
        if os.path.exists(path):
            with open(path, encoding="utf-8") as txt:
                for line in txt.readlines():
                    line = line.strip()
                    if line:
                        uploaded_parts.append(line.split(maxsplit=1)[1])
            client.logger.debug(
                "Uploads txt found with %s parts already uploaded", len(uploaded_parts)
            )
        parts = sorted(os.listdir(tempdir))
        pool = ThreadPoolExecutor(max_workers=max_workers)
        pbar = tqdm.tqdm(total=len(parts))
        with open(path, "a", encoding="utf-8") as txt:
            task = functools.partial(
                _upload_task,
                client=client,
                folder=tempdir,
                pbar=pbar,
                uploaded=uploaded_parts,
                txt_file=txt,
                lock=Lock(),
            )
            client.login()
            for _ in pool.map(task, parts):
                pbar.refresh()
    return path


def _upload_task(
    name: str,
    folder: str,
    uploaded: list,
    client: ToDusClient2,
    pbar: tqdm.tqdm,
    txt_file: TextIO,
    lock: Lock,
) -> None:
    path = os.path.join(folder, name)
    if name in uploaded:
        tqdm.tqdm.write(f"Skipping: {name}")
        os.remove(path)
        pbar.update(1)
        return
    tqdm.tqdm.write(f"Uploading: {name}")
    with open(path, "rb") as file:
        part = file.read()
    while True:
        try:
            url = client.upload_file(part, len(part))
            with lock:
                txt_file.write(f"{url}\t{name}\n")
                uploaded.append(name)
            os.remove(path)
            pbar.update(1)
            break
        except Exception as err:
            client.logger.exception(err)
            time.sleep(15)
            client.login()
            tqdm.tqdm.write(f"Retrying: {name} (ERROR: {err})")


def _get_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        prog=__name__.split(".", maxsplit=1)[0],
        description="ToDus Client",
    )
    parser.add_argument(
        "-n",
        "--number",
        dest="number",
        metavar="PHONE-NUMBER",
        default="",
        help="account's phone number, if not given the default account will be used",
    )
    parser.add_argument(
        "-v",
        "--version",
        action="version",
        version=__version__,
        help="show program's version number and exit.",
    )

    subparsers = parser.add_subparsers(dest="command")

    subparsers.add_parser(name="login", help="authenticate in server")

    up_parser = subparsers.add_parser(name="upload", help="upload file")
    up_parser.add_argument(
        "-s",
        "--split",
        dest="part_size",
        type=int,
        default=0,
        help="if given, the file will be split in parts of the given size (in bytes)",
    )
    up_parser.add_argument(
        "-w",
        "--max-workers",
        dest="max_workers",
        type=int,
        default=1,
        help="Number of simultaneous uploads (default: %(default)s)",
    )
    up_parser.add_argument("file", nargs="+", help="file to upload")

    down_parser = subparsers.add_parser(name="download", help="download file")
    down_parser.add_argument(
        "-w",
        "--max-workers",
        dest="max_workers",
        type=int,
        default=4,
        help="Number of simultaneous downloads (default: %(default)s)",
    )
    down_parser.add_argument("url", nargs="+", help="url to download or txt file path")

    subparsers.add_parser(name="token", help="get a token")

    acc_parser = subparsers.add_parser(name="accounts", help="list accounts")
    acc_parser.add_argument(
        "-r",
        "--remove",
        metavar="ACCOUNT",
        default="",
        help="Remove account from accounts list",
    )
    acc_parser.add_argument(
        "-d",
        "--default",
        metavar="ACCOUNT",
        default="",
        help="Set account as default account",
    )

    return parser


def _register(client: ToDusClient2, acc: dict, config: dict) -> None:
    if not client.phone_number:
        client.phone_number = normalize_phone_number(input("Enter Phone Number: "))
    client.request_code()
    pin = input("Enter PIN: ").strip()
    client.validate_code(pin)
    acc["phone_number"] = client.phone_number
    acc["password"] = client.password
    if acc not in config["accounts"]:
        config["accounts"].append(acc)
    _save_config(config)


def _get_password(phone: str, folder: str) -> str:
    path = os.path.join(folder, phone + ".cfg")
    if os.path.exists(path):
        with open(path, encoding="utf-8") as file:
            return file.read().split("=", maxsplit=1)[-1].strip()
    return ""


def _upload(client: ToDusClient2, args) -> None:
    for path in args.file:
        if args.part_size:
            tqdm.tqdm.write(f"Splitting: {path}")
            txt = _split_upload(client, path, args.part_size, args.max_workers)
            tqdm.tqdm.write(f"TXT: {txt}")
        else:
            tqdm.tqdm.write(f"Uploading: {path}")
            pbar = tqdm.tqdm(total=1)
            with open(path, "rb") as file:
                data = file.read()
            client.login()
            url = client.upload_file(data, len(data))
            pbar.update(1)
            pbar.refresh()
            url += "?name=" + quote_plus(os.path.basename(path))
            tqdm.tqdm.write(f"URL: {url}")


def _download(client: ToDusClient2, args) -> None:
    downloads = []
    for url in args.url:
        if url.startswith("http"):
            url, name = url.split("?name=", maxsplit=1)
            downloads.append((url, unquote_plus(name)))
        else:
            with open(url, encoding="utf-8") as file:
                for line in file.readlines():
                    line = line.strip()
                    if line:
                        url, name = line.split(maxsplit=1)
                        downloads.append((url, name))

    pool = ThreadPoolExecutor(max_workers=args.max_workers)
    pbar = tqdm.tqdm(total=len(downloads))
    task = functools.partial(_download_task, client=client, pbar=pbar)
    client.login()
    for _ in pool.map(task, downloads):
        pbar.refresh()


def _download_task(download: tuple, client: ToDusClient2, pbar: tqdm.tqdm) -> None:
    url, name = download
    url_display = url if len(url) < 50 else url[:50] + "..."
    if os.path.exists(name):
        tqdm.tqdm.write(f"Skipping: {name} ({url_display})")
        pbar.update(1)
        return
    tqdm.tqdm.write(f"Downloading: {name} ({url_display})")
    while True:
        try:
            client.download_file(url, name)
            pbar.update(1)
            break
        except Exception as err:
            client.logger.exception(err)
            time.sleep(15)
            client.login()
            tqdm.tqdm.write(f"Retrying: {name} (ERROR: {err})")


def _select_account(phone_number: str, config: dict) -> dict:
    if phone_number:
        phone_number = normalize_phone_number(phone_number)
        for acc in config["accounts"]:
            if acc["phone_number"] == phone_number:
                return acc
        return dict(phone_number=phone_number, password="")

    if config["accounts"]:
        acc = config["accounts"][0]
    else:
        acc = dict(phone_number="", password="")
    return acc


def _list_accounts(config: dict) -> None:
    if not config["accounts"]:
        print("No accounts added yet.")
    else:
        def_acc = config["accounts"][0]
        for acc in config["accounts"]:
            status = "logged" if acc["password"] else "not logged"
            print(
                f"{acc['phone_number']} ({status}){' [default]' if acc is def_acc else ''}"
            )


def main() -> None:
    """CLI program."""
    try:
        parser = _get_parser()
        args = parser.parse_args()

        config = _get_config()
        if args.command == "login":
            acc = dict(phone_number=args.number, password="")
        else:
            acc = _select_account(args.number, config)

        client = ToDusClient2(
            acc["phone_number"], acc["password"], logger=_get_logger()
        )
        if not client.registered and args.command not in ("", "login", "accounts"):
            print("ERROR: account not authenticated, login first.")
            return
        if args.command == "upload":
            _upload(client, args)
        elif args.command == "download":
            _download(client, args)
        elif args.command == "login":
            _register(client, acc, config)
        elif args.command == "token":
            client.login()
            print(client.token)
        elif args.command == "accounts":
            if args.remove:
                for acc in config["accounts"]:
                    if args.remove == acc["phone_number"]:
                        config["accounts"].remove(acc)
                        _save_config(config)
                        print(f"Account {acc['phone_number']!r} removed.")
                        break
                else:
                    print(f"ERROR: Account {args.remove!r} not found.")
            elif args.default:
                for acc in config["accounts"]:
                    if args.default == acc["phone_number"]:
                        config["accounts"].remove(acc)
                        config["accounts"].insert(0, acc)
                        _save_config(config)
                        print(f"Account {acc['phone_number']!r} set as default.")
                        break
                else:
                    print(f"ERROR: Account {args.default!r} not found.")
            else:
                _list_accounts(config)
        else:
            parser.print_usage()
    except AuthenticationError:
        acc["password"] = ""
        _save_config(config)
        print(f"ERROR: Session expired for account: {acc['phone_number']}")
    except KeyboardInterrupt:
        print("\nOperation canceled by user.")
        os._exit(1)  # noqa


PROGRAM_FOLDER = os.path.expanduser("~/.todus")
CONFIG_PATH = os.path.join(PROGRAM_FOLDER, "config.json")
if not os.path.exists(PROGRAM_FOLDER):
    os.makedirs(PROGRAM_FOLDER)
if not os.path.exists(CONFIG_PATH):
    _save_config(
        {
            "accounts": [],
        }
    )
